Buffer/scullBuffer.c

```diff
@@ -12,61 +12,65 @@ module_param(scull_size, int, S_IRUGO);
+struct scull_buffer scullBufferDevice;/* instance of the scullBuffer structure */
 
 /* file ops struct indicating which method is called for which io operation */
 struct file_operations scullBuffer_fops = {
-	.owner =    THIS_MODULE,
-	.read =     scullBuffer_read,
-	.write =    scullBuffer_write,
-	.open =     scullBuffer_open,
-	.release =  scullBuffer_release,
+  .owner =    THIS_MODULE,
+  .read =     scullBuffer_read,
+  .write =    scullBuffer_write,
+  .open =     scullBuffer_open,
+  .release =  scullBuffer_release,
 };
 
 /*
  * Method used to allocate resources and set things up when the module
  * is being loaded. 
-*/
+ */
 int scull_init_module(void)
 {
-	int result = 0;
-	dev_t dev = 0;
-	
-	/* first check if someone has passed a major number */
-	if (scull_major) {
-		dev = MKDEV(scull_major, scull_minor);
-		result = register_chrdev_region(dev, SCULL_NR_DEVS, "scullBuffer");
-	} else {
-		/* we need a dynamically allocated major number..*/
-		result = alloc_chrdev_region(&dev, scull_minor, SCULL_NR_DEVS,
-				"scullBuffer");
-		scull_major = MAJOR(dev);
-	}
-	if (result < 0) {
-		printk(KERN_WARNING "scullBuffer: can't get major %d\n", scull_major);
-		return result;
-	}
-	
-	/* alloc space for the buffer (scull_size bytes) */
-	scullBufferDevice.bufferPtr = kmalloc( scull_size , GFP_KERNEL);
-	if(scullBufferDevice.bufferPtr == NULL)
-	{
-		scull_cleanup_module();
-		return -ENOMEM;
-	}
-	
-	/* Init the count vars */
-	scullBufferDevice.readerCnt = 0;
-	scullBufferDevice.writerCnt = 0;
-	scullBufferDevice.size = 0;
-		
-	/* Initialize the semaphore*/
-	sema_init(&scullBufferDevice.sem, 1);
-	
-	/* Finally, set up the c dev. Now we can start accepting calls! */
-	scull_setup_cdev(&scullBufferDevice);
-	printk(KERN_DEBUG "scullBuffer: Done with init module ready for requests buffer size= %d\n",scull_size);
-	return result; 
+  int result = 0;
+  dev_t dev = 0;
+
+  /* first check if someone has passed a major number */
+  if (scull_major) {
+    dev = MKDEV(scull_major, scull_minor);
+    result = register_chrdev_region(dev, SCULL_NR_DEVS, "scullBuffer");
+  } else {
+    /* we need a dynamically allocated major number..*/
+    result = alloc_chrdev_region(&dev, scull_minor, SCULL_NR_DEVS,
+				 "scullBuffer");
+    scull_major = MAJOR(dev);
+  }
+  if (result < 0) {
+    printk(KERN_WARNING "scullBuffer: can't get major %d\n", scull_major);
+    return result;
+  }
+
+  /* alloc space for the buffer (scull_size bytes) */
+  scullBufferDevice.bufferPtr = kmalloc( scull_size*516 , GFP_KERNEL);
+  if(scullBufferDevice.bufferPtr == NULL)
+    {
+      scull_cleanup_module();
+      return -ENOMEM;
+    }
+
+  /* Init the count vars */
+  scullBufferDevice.readerCnt = 0;
+  scullBufferDevice.writerCnt = 0;
+  scullBufferDevice.front = 0;
+  scullBufferDevice.nextEmpty = 0;
+  scullBufferDevice.size = 0;
+
+  /* Initialize the semaphores*/
+  sema_init(&scullBufferDevice.sem, 1);
+  sema_init(&scullBufferDevice.empty, scull_size);
```

Comment:
I think empty should be initialized to 0 because at the beginning there is nothing to read.

Reply:
No it needs to be scull_size. empty is the semaphore that allows producers to write. If its 0 initially, nobody will be able to write. It basically means that NITEMS blocks are empty initially.

full is the semaphore for reading